Bugzilla/Install/AssetManager.pm

```diff
@@ -0,0 +1,224 @@
+use File::Copy qw(cp);
+use File::Find qw(find);
+use File::Basename qw(dirname);
+use File::Spec;
+use JSON::XS ();
+use MIME::Base64 qw(encode_base64);
+use File::Slurp;
+use List::MoreUtils qw(any all);
+use Carp;
+
+use Bugzilla::Constants qw(bz_locations);
+
+our $VERSION = 1;
+
+my $SHA_VERSION = '224';
+
+my $ABSOLUTE_DIR = declare as Str, 
+    where { File::Spec->file_name_is_absolute($_) && -d $_ }
+    message { "must be an absolute path to a directory" };
+
+has 'base_dir'       => ( is => 'lazy', isa => $ABSOLUTE_DIR );
+has 'asset_dir'      => ( is => 'lazy', isa => $ABSOLUTE_DIR );
+has 'source_dirs'    => ( is => 'lazy' );
+has 'state'          => ( is => 'lazy' );
+has 'state_file'     => ( is => 'lazy' );
+has 'json'           => ( is => 'lazy' );
+
+sub asset_file {
+    my ($self, $file, $relative_to) = @_;
+    $relative_to //= $self->base_dir;
+    my $asset_file = $self->state->{asset_map}->{$file}
+        or return $file;
+
+    return File::Spec->abs2rel(
+        File::Spec->catfile($self->asset_dir, $asset_file),
+        $relative_to
+    );
+}
+
+sub asset_files {
+    my ( $self, @files ) = @_;
+
+    # By default, this returns the files it was passed.
+    # They will be passed to asset_file() from the template.
+
+    return @files;
+}
+
+
+sub asset_sri {
+    my ($self, $asset_file) = @_;
+    my ($hex) = $asset_file =~ m!([[:xdigit:]]+)\.\w+$!;
+    my $data = pack "H*", $hex;
+    return "sha$SHA_VERSION-" . encode_base64($data, "");
+}
+
+sub compile_file {
+    my ($self, $file) = @_;
+    return unless -f $file;
+    my $base_dir  = $self->base_dir;
+    my $asset_dir = $self->asset_dir;
+    my $asset_map = $self->state->{asset_map};
+
+    my $key = File::Spec->abs2rel( $file, $base_dir );
+    return if $asset_map->{$key};
+
+    if ($file =~ /\.(jpe?g|png|gif|ico|woff|js)$/i) {
+        my $ext            = $1;
+        my $digest         = $self->_digest_file($file);
+        my $asset_file     = File::Spec->catfile($asset_dir, "$digest.$ext");
+        cp($file, $asset_file);
+        if ($digest eq $self->_digest_file($asset_file)) {
+            $asset_map->{$key} = File::Spec->abs2rel($asset_file, $asset_dir);
+        }
+        else {
+            die "failed to write $asset_file";
+        }
+    }
+    elsif ($file =~ /\.css$/) {
+        my $content = read_file($file);
+
+        # minify
+        $content =~ s{(?<!=)url\(([^\)]+)\)}{$self->_css_url_rewrite($1, $file)}eig;
+        my $digest = $self->_digest_string($content);
+        my $asset_file     = File::Spec->catfile($asset_dir, "$digest.css");
+        write_file($asset_file, $content);
+        if ($digest eq $self->_digest_file($asset_file)) {
+            $asset_map->{$key} = File::Spec->abs2rel($asset_file, $asset_dir);
+        }
+        else {
+            die "failed to write $asset_file";
+        }
+    }
+}
+
+sub compile_all {
+    my ($self) = @_;
+    my $asset_map = $self->state->{asset_map};
+
+    %$asset_map = ();
```

Comment:
That used to be the case when asset_map was itself an object attribute! Now that 'state' is the attribute, setting a fresh hashref is fine.